[6.2] [webservices] use logic and when filter multiple tag values

alikon opened the pull request:
Pull Request for Issue #46241 .

### Summary of Changes
This change adds support for a tag_mode filter to the Articles list in com_content so callers can choose whether tag filtering should use OR (any) or AND (all) semantics. It also exposes the new filter via the API controller.

What this does

1. Adds a new model state key filter.tag_mode (default: "any") to ArticlesModel.
2. Implements "all" tag mode: when tag_mode=all and multiple tag IDs are provided, an article must have all of the specified tags to be included (AND semantics).
3. Preserves the existing behavior as the default: tag_mode=any (OR semantics).
4. Exposes the tag_mode parameter in the API controller (ArticlesController::displayList) and cleans input with the request filter.


### Testing Instructions
 GET `api/index.php/v1/content/articles?filter[tag][]=2&filter[tag][]=3&filter[tag_mode]=all`

Manual tests performed for:
  - Single tag with default behavior (any)
  - Multiple tags with tag_mode=any (OR)
  - Multiple tags with tag_mode=all (AND)
  - Missing/invalid tag_mode falling back to default

### Actual result BEFORE applying this Pull Request

only logic OR

### Expected result AFTER applying this Pull Request
 Adds logic AND


### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [ ] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [ ] No documentation changes for manual.joomla.org needed

alikon committed:
logic-and

alikon committed:
cs

richard67 commented on administrator/components/com_content/src/Model/ArticlesModel.php:
```diff
@@ -147,6 +172,8 @@ protected function populateState($ordering = 'a.id', $direction = 'desc')
         $this->getUserStateFromRequest($this->context . '.filter.access', 'filter_access');
         $this->getUserStateFromRequest($this->context . '.filter.language', 'filter_language', '');
         $this->getUserStateFromRequest($this->context . '.filter.checked_out', 'filter_checked_out', '');
+        $this->getUserStateFromRequest($this->context . '.filter.tag_mode', 'filter_tag_mode', 'any');
+
```
```suggestion
        $this->getUserStateFromRequest($this->context . '.filter.tag_mode', 'filter_tag_mode', 'any');
```
Remove redundant empty line.

alikon replied:
done

alikon committed:
reduntant line

alikon committed:
Merge branch '6.1-dev' into webservices-logic-and-for-tags

alikon committed:
Merge branch '6.1-dev' into webservices-logic-and-for-tags

alikon committed:
Merge branch '6.1-dev' into webservices-logic-and-for-tags

alikon commented:
do you have some `{loadmoduleid xxx}` or similar  in one of your articles ?

alikon committed:
Merge branch '6.1-dev' into webservices-logic-and-for-tags

TerryHarker commented:
I have tested this item :white_check_mark: successfully on 3642879029b7f4a9c594fe903ee972942f1e93b4<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/46325">issues.joomla.org/tracker/joomla-cms/46325</a>.</sub>

alikon commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/46325">issues.joomla.org/tracker/joomla-cms/46325</a>.</sub>

jikubiswal commented:
I have tested this item :white_check_mark: successfully on 3642879029b7f4a9c594fe903ee972942f1e93b4<br /><br />I have verified that the patch for Applied Commit SHA: 3642879029
 is functioning correctly.
  I confirmed that ArticlesModel.php now correctly handles the tag_mode filter, switching to AND logic when tag_mode=all. I also confirmed that ArticlesController.php exposes this filter to the API.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/46325">issues.joomla.org/tracker/joomla-cms/46325</a>.</sub>

KishoriBKarale commented:
I have tested this item :white_check_mark: successfully on 3642879029b7f4a9c594fe903ee972942f1e93b4<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/46325">issues.joomla.org/tracker/joomla-cms/46325</a>.</sub>

richard67 committed:
Merge branch '6.1-dev' into webservices-logic-and-for-tags

richard67 commented on administrator/components/com_content/src/Model/ArticlesModel.php:
```diff
@@ -48,33 +48,58 @@ public function __construct($config = [], ?MVCFactoryInterface $factory = null)
+                'rating_count',
+                'rating',
+                'stage',
+                'wa.stage_id',
```
```suggestion
                'id', 'a.id',
                'title', 'a.title',
                'alias', 'a.alias',
                'checked_out', 'a.checked_out',
                'checked_out_time', 'a.checked_out_time',
                'catid', 'a.catid', 'category_title',
                'state', 'a.state',
                'access', 'a.access', 'access_level',
                'created', 'a.created',
                'modified', 'a.modified',
                'created_by', 'a.created_by',
                'created_by_alias', 'a.created_by_alias',
                'ordering', 'a.ordering',
                'featured', 'a.featured',
                'featured_up', 'fp.featured_up',
                'featured_down', 'fp.featured_down',
                'language', 'a.language',
                'hits', 'a.hits',
                'publish_up', 'a.publish_up',
                'publish_down', 'a.publish_down',
                'published', 'a.published',
                'author_id',
                'category_id',
                'level',
                'tag',
                'rating_count', 'rating',
                'stage', 'wa.stage_id',
```
Revert unrelated and unnecessary code style changes.

richard67 replied:
@alikon This one here, too, please.

richard67 replied:
P.S.: Human tests and so RTC would still be valid as all changes after that were only code style changes.

alikon replied:
done

richard67 commented on administrator/components/com_content/src/Model/ArticlesModel.php:
```diff
@@ -456,7 +482,7 @@ protected function getListQuery()
                         ? ' OR ' . $db->quoteName('a.created_by') . ' IN (' . implode(',', $query->bindArray($authorId)) . ')'
                         : '') .
-                ')');
+                    ')');
```
```suggestion
                ')');
```
Revert unrelated and wrong code style change.

richard67 commented on administrator/components/com_content/src/Model/ArticlesModel.php:
```diff
@@ -533,37 +559,57 @@ protected function getListQuery()
-            );
-
-            if ($includeNone) {
-                $subQuery2 = $db->createQuery()
+            $tagMode = $this->getState('filter.tag_mode', 'any');
+            if ($tagMode === 'all') {
```
```suggestion

            $tagMode = $this->getState('filter.tag_mode', 'any');

            if ($tagMode === 'all') {
```
Code style: Add empty lines after end and before start of control structures.

richard67 commented on administrator/components/com_content/src/Model/ArticlesModel.php:
```diff
@@ -533,37 +559,57 @@ protected function getListQuery()
+                        ->where(
+                            '(' . $db->quoteName('tagmap.content_item_id') . ' IS NOT NULL OR '
+                                . $db->quoteName('tagmap2.content_item_id') . ' IS NULL)'
+                        );
```
```suggestion
                    ->where(
                        '(' . $db->quoteName('tagmap.content_item_id') . ' IS NOT NULL OR '
                            . $db->quoteName('tagmap2.content_item_id') . ' IS NULL)'
                    );
```
Code style: Fix indentation level.

richard67 commented on administrator/components/com_content/src/Model/ArticlesModel.php:
```diff
@@ -580,20 +626,20 @@ protected function getListQuery()
                     $db->quoteName('tagmap.content_item_id') . ' = ' . $db->quoteName('a.id')
                 )
-                ->where($db->quoteName('tagmap.content_item_id') . ' IS NULL');
+                    ->where($db->quoteName('tagmap.content_item_id') . ' IS NULL');
```
```suggestion
                ->where($db->quoteName('tagmap.content_item_id') . ' IS NULL');
```
Revert unrelated and wrong code style change.

richard67 commented on administrator/components/com_content/src/Model/ArticlesModel.php:
```diff
@@ -580,20 +626,20 @@ protected function getListQuery()
+                            $db->quoteName('tagmap.type_alias') . ' = ' . $db->quote('com_content.article'),
+                        ]
+                    )
+                    ->bind(':tag', $tag, ParameterType::INTEGER);
```
```suggestion
                ->where(
                    [
                        $db->quoteName('tagmap.tag_id') . ' = :tag',
                        $db->quoteName('tagmap.type_alias') . ' = ' . $db->quote('com_content.article'),
                    ]
                )
                ->bind(':tag', $tag, ParameterType::INTEGER);
```
Revert unrelated and wrong code style change.

richard67 reviewed (changes requested):
Please apply my change suggestions to revert unrelated and unnecessary and partly wrong code style changes.

alikon committed:
Update administrator/components/com_content/src/Model/ArticlesModel.php

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

alikon committed:
Update administrator/components/com_content/src/Model/ArticlesModel.php

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

alikon committed:
Update administrator/components/com_content/src/Model/ArticlesModel.php

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

alikon committed:
Update administrator/components/com_content/src/Model/ArticlesModel.php

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

alikon committed:
Update administrator/components/com_content/src/Model/ArticlesModel.php

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

richard67 commented:
I have no idea why integration and unit tests are failing for Windows at the PHP installation, but it is not related to this PR.

richard67 commented on administrator/components/com_content/src/Model/ArticlesModel.php:
```diff
@@ -534,36 +560,58 @@ protected function getListQuery()
                 $includeNone = true;
             }
```
The logic to include "None" (i.e. articles with no tags assigned) into the query
```
$includeNone = false;

if (\in_array(0, $tag)) {
    $tag         = array_filter($tag);
    $includeNone = true;
}
```
can be moved from here to the `else` with the "OR" logic, i.e. between the lines `// OR logic:` and `$subQuery = $db->createQuery()`, so it lookes like this:
```
} else {
    // OR logic:
    $includeNone = false;

    if (\in_array(0, $tag)) {
        $tag         = array_filter($tag);
        $includeNone = true;
    }

    $subQuery = $db->createQuery()
...
```
The logic to include articles without any tags makes only sense with the "OR" logic, and your actual code changes correctly handle it in the "OR" part. Moving this part mentioned here would make it perfect.

alikon committed:
Update ArticlesModel.php

tecpromotion committed:
Merge branch '6.1-dev' into webservices-logic-and-for-tags

HLeithner commented on administrator/components/com_content/src/Model/ArticlesModel.php:
```diff
@@ -534,36 +535,58 @@ protected function getListQuery()
+                        ->select('DISTINCT ' . $db->quoteName('content_item_id'))
+                        ->from($db->quoteName('#__contentitem_tag_map'))
+                        ->where([
+                            $db->quoteName('tag_id') . ' = ' . $db->quote($tagId),
```
why don't you use prepared parameters here? You only need to bind the parameter to the $query object with `bindArray()` and use the result or by normal `bind()`

alikon replied:
done

alikon committed:
Merge branch '6.1-dev' into webservices-logic-and-for-tags

alikon committed:
Use parameter binding for tag_id in query

MacJoom committed:
Merge branch '6.2-dev' into webservices-logic-and-for-tags

alikon committed:
Merge branch '6.2-dev' into webservices-logic-and-for-tags

alikon committed:
Refactor tag filtering logic in ArticlesModel

alikon committed:
cs

alikon committed:
Add integration tests for article tag filtering

alikon committed:
Refactor database cleanup queries in tests

alikon committed:
Refactor articles tag filtering tests for clarity

alikon committed:
Refactor tag logic in ArticlesModel for better binding

Updated the tag handling logic to ensure articles have at least one selected tag. Improved query binding for tag IDs to avoid runtime issues.

alikon committed:
Add integration tests for content API tag filtering

alikon committed:
Refactor database cleanup queries in tests

alikon committed:
Remove unnecessary comments in ArticlesModel.php

Removed comments explaining the handling of tag IDs in the query.
